### PR TITLE
Rename mir::Rvalue to Op.

### PR DESCRIPTION
**Pull request by bcata6**
Originally suggested by @eddyb.

### CODE REVIEW

**Comment by eddyb** [comment]
cc @rust-lang/wg-mir-opt

File: src/librustc_codegen_ssa/mir/analyze.rs [hunk 1]
```diff
@@ -215,7 +215,7 @@ impl<'mir, 'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> Visitor<'tcx>
     fn visit_assign(
         &mut self,
         place: &mir::Place<'tcx>,
-        rvalue: &mir::Rvalue<'tcx>,
+        rvalue: &mir::Op<'tcx>,
```
Comment (eddyb) [inline src/librustc_codegen_ssa/mir/analyze.rs:218]: Probably the next thing to do here would be to rename `rvalue` variables.

File: src/librustc_codegen_ssa/mir/analyze.rs [hunk 2]
```diff
@@ -215,7 +215,7 @@ impl<'mir, 'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> Visitor<'tcx>
...
         location: Location,
     ) {
         debug!("visit_assign(place={:?}, rvalue={:?})", place, rvalue);
```
Comment (eddyb) [inline src/librustc_codegen_ssa/mir/analyze.rs:221]: Including debug messages, heh.

File: src/librustc_middle/mir/mod.rs [hunk 3]
```diff
@@ -2042,7 +2042,7 @@ impl<'tcx> Operand<'tcx> {
 /// Rvalues
```
Comment (oli-obk) [inline src/librustc_middle/mir/mod.rs:2042]: This documentation should also be adjusted

**Comment by oli-obk** [comment]
What does `Op` mean here? `Operation`? `Operand` makes no sense, we already have it.

File: src/librustc_mir/borrow_check/type_check/mod.rs [hunk 4]
```diff
@@ -2333,34 +2331,34 @@ impl<'a, 'tcx> TypeChecker<'a, 'tcx> {
...
+            | Op::UnaryOp(..)
+            | Op::Discriminant(..) => {}
         }
     }
 
     /// If this rvalue supports a user-given type annotation, then
     /// extract and return it. This represents the final type of the
     /// rvalue and will be unified with the inferred type.
-    fn rvalue_user_ty(&self, rvalue: &Rvalue<'tcx>) -> Option<UserTypeAnnotationIndex> {
+    fn rvalue_user_ty(&self, rvalue: &Op<'tcx>) -> Option<UserTypeAnnotationIndex> {
```
Comment (oli-obk) [inline src/librustc_mir/borrow_check/type_check/mod.rs:2347]: also loads of function that should get renamed, not just variables

**Comment by eddyb** [comment]
> `Operand` makes no sense, we already have it.
Yeah, operands are the arguments to an operation (or "operator" I guess).

We could even rename `Op::BinaryOp` to `Op::Binary`, which reads really nice to me.

**Comment by wesleywiser** [comment]
@eddyb What's the motivation for this? `Rvalue` is much more descriptive and easier to grep for.

File: src/librustc_mir/interpret/step.rs [hunk 5]
```diff
@@ -136,12 +136,12 @@ impl<'mir, 'tcx, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
     /// type writes its results directly into the memory specified by the place.
     pub fn eval_rvalue_into_place(
```
Comment (oli-obk) [inline src/librustc_mir/interpret/step.rs:137]: another function name that needs a rename

**Comment by oli-obk** [comment]
I'm not sure about the rename, but I'm not against it either. `Rvalue` was some jargon I had to learn for compilers, but beyond my understanding that it means "right hand side of assignment" I never pretended to understand all the around *values properly. 

`Operation` would be more descriptive, but quite long.

**Comment by eddyb** [comment]
A long time ago, we got rid of "lvalue" everywhere and "rvalue" everywhere outside MIR.
This has been a TODO ever since, although the replacement is definitely a matter of bikeshed.

Anyway one of the reasons I told @bcata6 to open this before spending more time on it is so we can make sure it's not a dead-end (i.e. it's not a big deal if we just close this PR).

**Comment by wesleywiser** [comment]
Got it, thanks! `Op` is a bit short for my taste, I think I would prefer if the whole word was spelled out. However, I don't feel strongly.

**Comment by eddyb** [comment]
Alright, I'm cc-ing the broader @rust-lang/compiler, and nominating this as a bikeshed item.

**Comment by wesleywiser** [comment]
I think before merging this, we should also try to enumerate the other places that will need to be updated with this change and have a plan for updating those as well. For example,

- various methods with `rvalue` in the name as @oli-obk  has already pointed out
- doc comments mentioning `rvalue`s
- rustc dev guide has plenty of `Rvalue` mentions I'm sure
- miri
- cg_cranelift (would be nice to leave a PR for @bjorn3 so they don't have to deal with this)
- others?

**Comment by Centril** [comment]
> others?

The reference and the UCG.

I personally think "rvalue" is fairly established terminology by now, and `Op` vs. `Operand` makes it harder to grep for, so I don't find the churn here useful.

**Comment by eddyb** [comment]
> I personally think "rvalue" is fairly established terminology by now

It's no more established than it was at the time of e.g. https://github.com/rust-lang/rust-by-example/issues/1028 (sadly, it looks like we didn't track this properly, which explains how we lost sight of `mir::Rvalue`).

**Comment by pnkfelix** [comment]
discussed in ad hoc [triage](https://zulip-archive.rust-lang.org/131828tcompiler/35095adhoctriagemeeting2020041054818.html#193574805) after friday planning meeting.

my inclination is to close this. I don't think `Op` is an improvement over `Rvalue`. We can separately bikeshed potential alternatives.

@rfcbot fcp close

**Comment by crlf0710** [comment]
I think exactly this bike-shedding [has happened before in wg-ucg](https://github.com/rust-lang/unsafe-code-guidelines/pull/40), cc @RalfJung 

Personally i prefer something like `ValExpr` here (if "expr" part is not good, maybe replace it with something like "form"), and i really wish the "left/right" terminology be entirely removed from Rust for consistency.

**Comment by eddyb** [comment]
> Personally i prefer something like `ValExpr` here (if "expr" part is not good, maybe replace it with something like "form")

`ValOp` and `ValKind` are two quick ideas based on that.

**Comment by petrochenkov** [comment]
I mostly have no idea about what's happening here, but I've checked my box because while name `Rvalue` could give some intuition about its meaning, the name `Op` can mean absolutely anything.

**Comment by RalfJung** [comment]
> I think exactly this bike-shedding has happened before in wg-ucg,

Indeed, and `Op` is a bad name here I think. We already have `Operand`, so this basically means we have two distinct concepts with the same name.

`Value` would be in accordance with the conclusion of the UCG bikeshedding and consistent with `Place`.

I like `ValExpr` (and it is also consistent with UCG bikeshedding) but then it also should be `PlaceExpr` (and `OpExpr`, maybe).

> i really wish the "left/right" terminology be entirely removed from Rust for consistency

Agreed.

**Comment by eddyb** [comment]
> Indeed, and `Op` is a bad name here I think. We already have `Operand`, so this basically means we have two distinct concepts with the same name.

I still think "operand" are what "operators"/"operations" *operate on*. AM I missing something?

**Comment by RalfJung** [comment]
> I still think "operands" are what "operators"/"operations" operate on. Am I missing something?

I mean yes, but that's not helpful here. By that standard `mir::StatementKind::Assign` has two operands, one on the left and one on the right, which are the two things the assignment operates on.

The way I view it, the fundamental notions are *places* and *values*. This is [reflected by the reference](https://doc.rust-lang.org/reference/expressions.html#place-expressions-and-value-expressions). Assignment puts a value into a place.

`mir::Operand` is an implementation detail, which for simplicity and effieiciency carves out a subset of values that can be directly passed to primitive operations and functions -- but in principle, at least as far as a MIR spec is concerned, there should be no problem at all in supporting arbitrary MIR value expressions (`mir::Rvalue`) anywhere that we currently use `mir::Operand`. (Miri exploits the simpler nature of `mir::Operand` for its "lazy" implementation of copying, but that should be an unobservable implementation detail of the interpreter, not something that exists in the MIR spec.)

**Comment by nikomatsakis** [comment]
For what it's worth, I have found the term "place" to be "not great" when talking to people, particularly when it's referring (ultimately) to a *place expression*. I spend a while thinking about this in the polonius talk and I landed on "path expression" (or frequently just "path") as the term for "place expression", because it seemed like whenever I would try to explain what a "place" is, I always explained it by saying "it's a path, like a.b.c", and people understood that. 

In short, I think I largely agree with @RalfJung's proposed nomenclature of places and values as things that occur at runtime.

I could live with place/value expressions as the thing we manipulate in our MIR programs (that will be evaluated to particular places/values), but actually I think the symmetry isn't that great, because it makes the terms rather verbose, and I think people won't want to use them. `PlaceExpr` and `PlaceExpression` are long and unwieldy.

Personally, I prefer place/path -- i.e., what we call `Place` in MIR we would call `Path`, and a place would be the dynamic stack slot or whatever that a path evaluates to.

I was going to say I'd prefer value/expression, but I realize I can't really decide whether I think that expression should an `Rvalue` or an `Operand` =) Probably `Rvalue`. 

Maybe we ought to try and draw up a big table with the options... or maybe introducing place/path is not helpful here. But it would be good to settle these names (and align them with UCG terminology).

**Comment by RalfJung** [comment]
> Personally, I prefer place/path -- i.e., what we call Place in MIR we would call Path, and a place would be the dynamic stack slot or whatever that a path evaluates to.

I feel rather strongly that it would be strange to call the dynamic concept a place and the static concept a path -- that sounds like they are unrelated. It should be "path expression" and "path value" or so.

> but I realize I can't really decide whether I think that expression should an Rvalue or an Operand =) Probably Rvalue.

Hm, I thought we had consensus for "value expression" in the UCG discussion, and that's also what the reference uses, for whatever that is worth.

**Comment by nikomatsakis** [comment]
Heh, I'm probably just wrong -- or at least "unique" -- in finding four distinct, short words more memorable and evocative than compound terms like "place expression". I can certainly see that having the symmetry could help people trying to get their head around the distinctions at play. I still fear that the terms "place expression" and "value expression" are unwieldy and people will tend not to use them in full, but I will happily yield to the UCG consensus and stop trying to derail. =)

But I do have a point of clarification...

> Hm, I thought we had consensus for "value expression" in the UCG discussion, and that's also what the reference uses, for whatever that is worth.

But what does it use it to refer *to* -- what we call "rvalues" here or "operands"? Given your point (which I agree with) that the [a-normalized form](http://matt.might.net/articles/a-normalization/) of MIR is a convenient implementation detail and not something "fundamental", I would expect that "value expressions" refers to "rvalues", and that *operands* are a kind of subset of that which we happen to find convenient (and hence they could, for example, be called "value subexpressions" or something like that, too).

**Comment by RalfJung** [comment]
> But what does it use it to refer to -- what we call "rvalues" here or "operands"? Given your point (which I agree with) that the a-normalized form of MIR is a convenient implementation detail and not something "fundamental", I would expect that "value expressions" refers to "rvalues",

Yes that is my understanding, too. Hence the suggestion to rename "rvalue" to "value" in this PR.

> and that operands are a kind of subset of that which we happen to find convenient (and hence they could, for example, be called "value subexpressions" or something like that, too).

Operands are a subset of values, but a "subexpression" is [something else](https://chortle.ccsu.edu/Java5/Notes/chap09B/ch09B_10.html). I think "operand" is not bad, if I had to come up with a name I might have called it a "leaf value expression" or so (since operands are the leaves of value expressions).

**Comment by nikomatsakis** [comment]
For the record, I'm fine with the names 

* `Rvalue => Value`
* `Operand => Operand`

That said, from that link you sent, @RalfJung:

> A subexpression is a part of an expression that is by itself a correct expression. 

In what way is an operand not itself a correct expression? I think that's what the [`Rvalue::Use`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/enum.Rvalue.html#variant.Use) variant is -- basically "upcasting" an `Operand` into an `Rvalue`.

**Comment by RalfJung** [comment]
Renaming "operand" to "subexpression" is not even "well-typed" in some ontological sense... there is no such thing as a "subexpression" in a global way, there are just "subexpressions of some given expression E". In other way, "subexpression" is a binary relation on expressions (it has type `expr -> expr -> Prop`). But "operand" is just a (sub)set of expressions, it has type `expr -> Prop`.

> In what way is an operand not itself a correct expression?

Uh, it is, but so what? The defining property of an operand is not "this is a subexpression of some `e`". The defining property is "this is an expression that satisfies these additional constraints".

Renaming "operand" to "subexpression" is a bit like renaming "square" to "subclass" because, you know, "square" is a subclass of "shape". (Not a great analogy, somehow I cannot come up with a better one right now.)

**Comment by spastorino** [comment]
Unnominating as it was discussed already in our last meeting.

**Comment by spastorino** [comment]
Renominated with the purpose of mentioning in the meeting that the discussion is still ongoing.

**Comment by pnkfelix** [comment]
Okay I'm officially moving the bikeshed regarding the name over to Zulip.

Here is the Zulip thread for the bikeshed: https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/To.20what.20to.20rename.20mir.3A.3ARvalue.20.2370928/near/195072985

I think we are all agreed that we won't land the renaming as encoded in this PR, so I am going to go ahead and close this PR. Thank you very much for the PR, @bcata6 !

**Comment by RalfJung** [comment]
So my understanding is that if a PR was opened that renamed `Rvalue` to `Value`, that would be accepted? Or is more discussion needed first?

**Pull request closed without merging**